homeassistant/components/vizio/config_flow.py

```diff
@@ -0,0 +1,161 @@
+        """Initialize config flow."""
+        self.import_schema = None
+        self.user_schema = None
+
+    async def async_step_user(
+        self, user_input: Dict[str, Any] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if user_input is not None:
+            # Store current values in case setup fails and user needs to edit
+            self.user_schema = update_schema_defaults(user_input)
+
+            # Check if new config entry matches any existing config entries
+            for entry in self.hass.config_entries.async_entries(DOMAIN):
+                if entry.data[CONF_HOST] == user_input[CONF_HOST]:
+                    errors[CONF_HOST] = "host_exists"
+                    break
+
+                if entry.data[CONF_NAME] == user_input[CONF_NAME]:
+                    errors[CONF_NAME] = "name_exists"
+                    break
+
+            if not errors:
+                try:
+                    # Ensure schema passes custom validation, otherwise catch exception and add error
+                    validate_auth(user_input)
+
+                    # Ensure config is valid for a device
+                    if not await VizioAsync.validate_ha_config(
+                        user_input[CONF_HOST],
+                        user_input.get(CONF_ACCESS_TOKEN),
+                        user_input[CONF_DEVICE_CLASS],
+                    ):
+                        errors["base"] = "cant_connect"
+                except vol.Invalid:
+                    errors["base"] = "tv_needs_token"
+
+            if not errors:
+                unique_id = await VizioAsync.get_unique_id(
+                    user_input[CONF_HOST],
+                    user_input.get(CONF_ACCESS_TOKEN),
+                    user_input[CONF_DEVICE_CLASS],
+                )
+
+                # Abort flow if existing component with same unique ID matches new config entry
+                if await self.async_set_unique_id(
+                    unique_id=unique_id, raise_on_progress=True
+                ):
+                    return self.async_abort(reason="already_setup")
+
+                return self.async_create_entry(
+                    title=user_input[CONF_NAME], data=user_input
+                )
+
+        schema = self.user_schema or self.import_schema or update_schema_defaults({})
+
+        return self.async_show_form(step_id="user", data_schema=schema, errors=errors)
+
+    async def async_step_import(self, import_config: Dict[str, Any]) -> Dict[str, Any]:
+        """Import a config entry from configuration.yaml."""
+        # Check if new config entry matches any existing config entries
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
+            if entry.data[CONF_HOST] == import_config[CONF_HOST] and entry.data[
+                CONF_NAME
+            ] == import_config.get(CONF_NAME):
+                return self.async_abort(reason="already_setup")
```

Comment:
If you want to allow users to update an existing config entry that was imported from config yaml, where eg the volume option has changed in config yaml compared to the existing config entry data, we could compare the new yaml data with the existing entry data and if there's a diff update the config entry data and config entry options here before we abort.

This way we allow users to update via config yaml but we also make sure that they can change via GUI options without overwriting such a change on every restart.

This could also be a later improvement.

Reply:
I will note this as a change to make later. I have two more incoming PRs, one to change the device type `soundbar` to `speaker`, and one to add `zeroconf` discovery support. I will add this to the `zeroconf` discovery PR

Reply:
Let me make this change now before I forget